modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/agg/Accumulators.java

```diff
@@ -541,27 +619,35 @@ public RelDataType returnType(IgniteTypeFactory typeFactory) {
         }
     }
 
-    private static class DecimalSumEmptyIsZero implements Accumulator {
+    /** SUM(DECIMAL) accumulator. */
+    public static class DecimalSumEmptyIsZero implements Accumulator {
         public static final Supplier<Accumulator> FACTORY = DecimalSumEmptyIsZero::new;
 
-        private BigDecimal sum = BigDecimal.ZERO;
-
         /** {@inheritDoc} */
         @Override
-        public void add(Object... args) {
+        public void add(AccumulatorsState state, Object... args) {
             BigDecimal in = (BigDecimal) args[0];
 
             if (in == null) {
                 return;
             }
 
-            sum = sum.add(in);
+            BigDecimal sum = (BigDecimal) state.get();
+            if (sum == null) {
+                state.set(in);
+            } else {
+                state.set(sum.add(in));
+            }
         }
 
         /** {@inheritDoc} */
         @Override
-        public Object end() {
-            return sum;
+        public void end(AccumulatorsState state, AccumulatorsState result) {
+            if (!state.hasValue()) {
```

Comment:
can we always put ZERO during creation of the accumulator and reduce all code that check have we null or not?

Reply:
This requires 3 replacing Sum(XXSumEmptyIsZero) with SumXXX aggregates.